File: src/main/java/net/benwoodworth/fastcraft/core/api/dependencies/event/EventListener.java
package net.benwoodworth.fastcraft.core.api.dependencies.event;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.function.Consumer;

/**
 * Listens to events, and notifies event handlers when events occur.
 *
 * @param <TEvent> the type of event being listened for
 */
public class EventListener<TEvent> {

    /**
     * This listener's event handlers.
     */
    @NotNull
    private ArrayList<Consumer<TEvent>> handlers = new ArrayList<>();

    /**
     * Raise an event.
     *
     * @param event the event to handle
     */
    public void notifyHandlers(@NotNull TEvent event) {
        handlers.forEach(handler -> handler.accept(event));
    }

    /**
     * Add an event handler to this listener.
     *
     * @param handler the event handler to add to this event
     */
    public void addHandler(@NotNull Consumer<TEvent> handler) {
        handlers.add(handler);
    }

    /**
     * Remove an event handler from this listener.
     *
     * @param handler the event handler to remove from this listener
     */
    public void removeHandler(@NotNull Consumer<TEvent> handler) {
        handlers.remove(handler);
    }
}

File: src/main/kotlin/net/benwoodworth/fastcraft/dependencies/event/EventListener.kt
package net.benwoodworth.fastcraft.dependencies.event

import java.util.*

/**
 * Listens to events, and notifies event handlers when events occur.
 *
 * @param TEvent the type of event being listened for
 */
class EventListener<TEvent> {

    /** This listener's event handlers. */
    private val handlers = HashSet<(event: TEvent) -> Unit>()

    /**
     * Raise an event.
     *
     * @param event the event to handle
     */
    fun notifyHandlers(event: TEvent) {
        handlers.forEach { it(event) }
    }

    /**
     * Add an event handler to this listener.
     *
     * @param handler the event handler to add to this event
     */
    fun addHandler(handler: (event: TEvent) -> Unit) {
        handlers += handler
    }

    /**
     * Remove an event handler from this listener.
     *
     * @param handler the event handler to remove from this listener
     */
    fun removeHandler(handler: (event: TEvent) -> Unit) {
        handlers -= handler
    }
}
